_source/_posts/2020-12-16-react-login.md

```diff
@@ -0,0 +1,599 @@
+        </label>
+        <button className="button">Login</button>
+      </form>
+    );
+  }
+};
+
+export default withOktaAuth(OktaSignInWidget);
+```
+
+You will notice a number of differences compared to the implementation in the previous section. First, the `render()` function now returns its own form with **Email** and **Password** input elements. The functions `handleEmailChange()` and `handlePasswordChange()` update the components state, always reflecting the value in the input fields.
+
+In the constructor, you will see that an `OktaAuth` object is created and assigned to the `oktaAuth` property. This object is used inside the `signIn()` function, which is called when the form is submitted. `OktaAuth` is responsible for connecting to the Okta servers, sending the user credentials, and in return, receiving a session token.
+
+The last thing you will notice is the fact that the `OktaSignInWidget` class is wrapped in the `withOktaAuth()` function before being exported. This adds the `authService` to the properties of the object. Remember, this widget is rendered within the `Security` widget. Inside the `signIn()` function the `authService` is used to set the application's session token and redirect the user to the protected page they wanted to visit before the sign-in was triggered.
+
+To round things up, add a bit of styling for the login form. Open `src/App.css` and paste the following code at the end of the file.
+
+```css
+.login-form {
+  display: inline-block;
+  width: 20em;
+}
+
+.full-width-input {
+  display: block;
+  text-align: right;
+}
+
+.full-width-input input {
+  width: 12em;
+  padding: 4px 8px;
+  margin: 4px 8px;
+  border: 1px solid #aaaaaa;
+  border-radius: 0;
+}
+```
+
+Start the application again by running `npm start`. Now, try to navigate to the protected page and you should see a login form that looks like the image below.
+
+{% img blog/react-login/react-self-hosted-login.png alt:"The self-hosted sign-in form" width:"800" %}{: .center-image }
+
+Congratulations! If you have made it this far, you now have the knowledge to implement the right sign-in solution for your React application.
+
+## Learn More about React and Authentication
+
+In this tutorial, I have shown you three ways to implement authentication in your React application. The easiest of these three options redirects the user to a hosted sign-in page on the Okta servers. While this is a perfectly good solution in many situations, you sometimes want to host the login form on your own servers. The Okta sign-in widget lets you embed the trusted login form on your own site. If you want a sign-in form with your layout and design, you can use the `OktaAuth` class to send the username and password directly to the Okta servers for validation.
+
+You can find the source code for this example in [the oktadeveloper/okta-react-login-example repository](https://github.com/oktadeveloper/okta-react-login-example) on GitHub.
```

Comment:
Will you eventually be updating the README of this repo to link back to this blog post?

Reply:
Yes, I just did!